src/components/ChannelList/ChannelList.js

```diff
@@ -486,58 +131,196 @@ class ChannelList extends PureComponent {
+   * 3. [ChannelPreviewMessanger](https://getstream.github.io/stream-chat-react/#ChannelPreviewMessanger)
+   *
+   * The Preview to use, defaults to ChannelPreviewLastMessage
+   * */
+  Preview: PropTypes.elementType,
+
+  /**
+   * Loading indicator UI Component. It will be displayed until the channels are
+   * being queried from API. Once the channels are loaded/queried, loading indicator is removed
+   * and replaced with children of the Channel component.
+   *
+   * Defaults to and accepts same props as:
+   * [LoadingChannels](https://github.com/GetStream/stream-chat-react/blob/master/src/components/LoadingChannels.js)
+   *
+   */
+  LoadingIndicator: PropTypes.elementType,
+  /**
+   * Error indicator UI Component. It will be displayed if there is any error if loading the channels.
+   * This error could be related to network or failing API.
+   *
+   * Defaults to and accepts same props as:
+   * [ChatDown](https://github.com/GetStream/stream-chat-react/blob/master/src/components/ChatDown.js)
+   *
+   */
+  LoadingErrorIndicator: PropTypes.elementType,
+  /**
+   * Custom UI Component for container of list of channels. Note that, list (UI component) of channels is passed
+   * to this component as children. This component is for the purpose of adding header to channel list or styling container
+   * for list of channels.
+   *
+   * Available built-in options (also accepts the same props as):
+   *
+   * 1. [ChannelListTeam](https://github.com/GetStream/stream-chat-react/blob/master/src/components/ChannelListTeam.js) (default)
+   * 2. [ChannelListMessenger](https://github.com/GetStream/stream-chat-react/blob/master/src/components/ChannelListMessenger.js)
+   *
+   * It has access to some additional props:
+   *
+   * - `setActiveChannel` {function} Check [chat context](https://getstream.github.io/stream-chat-react/#chat)
+   * - `activeChannel` Currently active channel object
+   * - `channels` {array} List of channels in channel list
+   */
+  List: PropTypes.elementType,
+  /**
+   * Paginator component for channels. It contains all the pagination logic such as
+   * - fetching next page of results when needed e.g., when scroll reaches the end of list
+   * - UI to display loading indicator when next page is being loaded
+   * - call to action button to trigger loading of next page.
+   *
+   * Available built-in options (also accepts the same props as):
+   *
+   * 1. [LoadMorePaginator](https://github.com/GetStream/stream-chat-react/blob/master/src/components/LoadMorePaginator.js)
+   * 2. [InfiniteScrollPaginator](https://github.com/GetStream/stream-chat-react/blob/master/src/components/InfiniteScrollPaginator.js)
+   */
+  Paginator: PropTypes.elementType,
+  /**
+   * Function that overrides default behaviour when new message is received on channel that is not being watched
+   *
+   * @param {Component} setChannels Setter for channels value in state.
```

Comment:
This will be a breaking change. So far we used to pass the `this` reference of component to these event handlers. But with functional component, that won't be possible. So instead we will pass `setChannels` function so channels can be altered from these handlers.

@jaapbakker88 @mahboubii @ath92 @vicnicius any comments?

Reply:
I think it makes sense, this way. let's just create a minor release when we do.